### PR TITLE
write filters section

### PR DESCRIPTION
**Pull request by Oblarg**
Adds a section on filters (and their WPILib classes) to advanced-controls

### CODE REVIEW

File: source/docs/software/advanced-control/filters/linear-filter.rst [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    LinearFilter filter = LinearFilter.highPass(0.1, 0.02);
+
+  .. code-tab:: c++
+
+    // Creates a new high-pass IIR filter
+    // Time constant is 0.1 seconds
+    // Period is 0.02 seconds - this is the standard FRC main loop period
+    frc::LinearFilter<double> filter = frc::LinearFilter<double>::HighPass(0.1_s, 0.02_s);
+
+The "time constant" parameter determines the "characteristic timescale" of the filter's impulse response; the filter will cancel out any signal dynamics that occur on timescales significantly longer than this.  Relatedly, it is also the approximate timescale of the introduced :ref:`phase lead <docs/software/advanced-control/filters/introduction:Phase lag>`.  The reciprocal of this timescale is the "cutoff frequency" of the filter.
```
Comment (calcmogul) [inline source/docs/software/advanced-control/filters/linear-filter.rst]: The time constant tau is actually 1/(2*pi*cuttoff frequency).

File: source/docs/software/advanced-control/filters/median-filter.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
+Median Filter
+=============
+
+A `robust <https://en.wikipedia.org/wiki/Robust_statistics>`__ alternative to the :ref:`moving-average filter <docs/software/advanced-control/filters/linear-filter:movingAverage>` is the *median filter*.  Where a moving average filter takes the arithmetic *mean* of the input over a moving sample window, a median filter (per the name) takes a median instead.
+
+The median filter is most-useful for removing occasional outliers from an input stream.  This makes it particularly well-suited to filtering inputs from distance sensors, which tend to be prone to occasional interference.  Unlike a moving average, the median filter will remain completely unaffected by small numbers of outliers, no matter how extreme.
```
Comment (calcmogul) [inline source/docs/software/advanced-control/filters/median-filter.rst]:
```suggestion
The median filter is most-useful for removing occasional outliers from an input stream.  This makes it particularly well-suited to filtering inputs from distance sensors, which are prone to occasional interference.  Unlike a moving average, the median filter will remain completely unaffected by small numbers of outliers, no matter how extreme.
```

**Comment by AustinShalit** [comment]
I think we need some graphs with that plot input and output signals to each filter to show how they behave

**Comment by Oblarg** [comment]
That's a good idea; I'll need a good choice of input signal, though.  Thoughts?

**Comment by sciencewhiz** [comment]
Fixes #359

**Comment by AustinShalit** [comment]
I do not. Maybe @calcmogul does?

**Comment by sciencewhiz** [comment]
Here's data from a noisy Sharp IR analog distance sensor. 
[SharpIRDistance.log](https://github.com/wpilibsuite/frc-docs/files/4075303/SharpIRDistance.log)

File: source/docs/software/advanced-control/filters/introduction.rst [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+So, what exactly *is* a filter, then?  Simply put, a filter is a mapping from a stream of inputs to a stream of outputs.  That is to say, the value output by a filter (in principle) can depend not only on the *current* value of the input, but on *the entire set of past and future values* (of course, in practice, the filters provided by WPILib are implementable in real-time on streaming data; accordingly, they can only depend on the *past* values of the input, and not on future values).  This is an important concept, because generally we use filters to remove/mitigate unwanted *dynamics* from a signal.  When we filter a signal, we're interested in modifying *how the signal changes over time*.
+
+Effects of Using a Filter
+-------------------------
+
+Noise Reduction
+^^^^^^^^^^^^^^^
+
+One of the most typical uses of a filter is for noise reduction.  A filter that reduces noise is called a *low-pass* filter (because it allows low frequencies to "pass through," while blocking high-frequencies).  Most of the filters currently included in WPIlib are effectively low-pass filters.
```
Comment (prateekma) [inline source/docs/software/advanced-control/filters/introduction.rst]: Capitalize `L` in `WPILib` near end of paragraph.

**Comment by Oblarg** [comment]
I've added plots.  I think they're pretty nifty.

**Comment by sciencewhiz** [comment]
Are the plots using the same values as the code examples? 

It would be helpful to include the data, both so that students could play with it, and so that if we add a new filter in the future we can make a similar image for the new filter. You can use the download rst tag for the data.

**Comment by Oblarg** [comment]
Yes, the plots are using the same values as the code examples.  I'll add the data for the plots, as well.

**Comment by sciencewhiz** [comment]
The Median filter is using 10 samples (0.2s) while everything else is 5 samples or a 0.1s time constant. For comparability, it would be best if they were the same.

**Comment by Oblarg** [comment]
That was a typo; the graph was indeed using 5 taps.  It's been fixed now.

File: source/docs/software/advanced-control/filters/index.rst [hunk 4]
```diff
@@ -0,0 +1,16 @@
+Filters
+=======
+
+.. note:: The data used to generate the various demonstration plots in this section can be found :download:`here <resources/filterdemo.csv>`.
+
+This section describes a number of filters included with WPILib that are useful for noise reduction and/or input smoothing.
+
+The data used to generate the demonstration plots for the various filters in this section can be found here:
```
Comment (sciencewhiz) [inline source/docs/software/advanced-control/filters/index.rst]: This probably should be deleted as it's OBE by the note above

**Pull request merged**